test/integration/job/job_test.go

```diff
@@ -1026,6 +1029,202 @@ func TestIndexedJob(t *testing.T) {
+		t.Run(name, func(t *testing.T) {
+			defer featuregatetesting.SetFeatureGateDuringTest(t, feature.DefaultFeatureGate, features.ElasticIndexedJob, tc.featureGate)()
+
+			closeFn, restConfig, clientSet, ns := setup(t, "indexed")
+			defer closeFn()
+			ctx, cancel := startJobControllerAndWaitForCaches(restConfig)
+			defer cancel()
+			resetMetrics()
+
+			// Set up initial Job in Indexed completion mode.
+			mode := batchv1.IndexedCompletion
+			jobObj, err := createJobWithDefaults(ctx, clientSet, ns.Name, &batchv1.Job{
+				Spec: batchv1.JobSpec{
+					Parallelism:    pointer.Int32Ptr(int32(initialCompletions)),
+					Completions:    pointer.Int32Ptr(int32(initialCompletions)),
+					CompletionMode: &mode,
+				},
+			})
+			if err != nil {
+				t.Fatalf("Failed to create Job: %v", err)
+			}
+			jobClient := clientSet.BatchV1().Jobs(jobObj.Namespace)
+
+			// Wait for pods to start up.
+			err = wait.PollImmediate(5*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+				job, err := jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
+				if err != nil {
+					return false, err
+				}
+				if job.Status.Active == int32(initialCompletions) {
+					return true, nil
+				}
+				return false, nil
+			})
+			if err != nil {
+				t.Fatalf("Error waiting for Job pods to become active: %v", err)
+			}
+
+			currentCompletions := initialCompletions
+
+			for _, update := range tc.jobUpdates {
+				// Update Job spec if necessary.
+				if update.completions != NoCompletionsUpdate {
+					if jobObj, err = updateJob(ctx, jobClient, jobObj.Name, func(j *batchv1.Job) {
+						j.Spec.Completions = pointer.Int32Ptr(int32(update.completions))
+						j.Spec.Parallelism = pointer.Int32Ptr(int32(update.completions))
+					}); err != nil {
+						// Error expected if feature gate is off
+						if !tc.featureGate {
+							return
+						}
+						t.Fatalf("Failed to update Job: %v", err)
+					}
+					currentCompletions = update.completions
+				}
+
+				// Succeed specified indexes.
+				succeededSet := sets.NewInt()
+				if update.succeededIndexes != nil {
+					for _, idx := range update.succeededIndexes {
+						if err := setJobPhaseForIndex(ctx, clientSet, jobObj, v1.PodSucceeded, idx); err != nil {
+							t.Fatalf("Failed trying to succeed pod with index %d: %v", idx, err)
+						}
+						succeededSet.Insert(idx)
+					}
+					newActive := currentCompletions - len(update.succeededIndexes)
+
+					// initialCompletions == initial parallelism, and active must be <= parallelism.
+					if newActive > currentCompletions && currentCompletions != NoCompletionsUpdate {
+						newActive = currentCompletions
+					}
+					validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+						Active:    newActive,
+						Succeeded: len(update.succeededIndexes),
+						Ready:     pointer.Int32(0),
+					})
+					validateIndexedJobPods(ctx, t, clientSet, jobObj, remainingIndexSet(currentCompletions, succeededSet), update.wantSucceededIndexes)
```

Comment:
We need to check for `wantSucceededIndexes` all the time, not only when `succeededIndexes` is set.

Reply:
otherwise, we are not properly verifying the test starting at line 1089 that the previously succeeded index was truly discarded.

Reply:
Ah good catch. Done.